### PR TITLE
Handle private properties in the compiled model

### LINKED ISSUE #24900: Compiled model: Use static binding instead of reflection for properties and fields

### LINKED ISSUE #24904: Compiled model: Generate lambdas used in change tracking
Depends on #24900 and #24895

### LINKED ISSUE #29761: Make change tracking and the update pipeline compatible with AOT/trimming
Note that although ExecuteUpdate/Delete allow bypassing change tracking and the update pipeline, we have no ExecuteInsert.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Add full support for service properties
Use a consistent pattern for unsafe accessors
Store the unsafe accessor names in an annotation
Refactor the different mechanisms for tracking variables in the generated code into a single bidirectional dictionary

Fixes #29761
Fixes #24904
Fixes #24900

### CODE REVIEW

File: src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs [hunk 1]
```diff
@@ -411,6 +413,39 @@ Expression VisitAssignment(BinaryExpression assignment, SyntaxKind kind)
...
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public static string GetUnsafeAccessorName(MemberInfo member)
+    {
+        StringBuilder stringBuilder = new();
+        stringBuilder.Clear().Append("UnsafeAccessor_");
```
Comment (ajcvickers) [inline src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs:425]: Is this our naming convention, or a convention recommended from somewhere else?

File: src/EFCore.Design/Query/Internal/RuntimeModelLinqToCSharpSyntaxTranslator.cs [hunk 2]
```diff
@@ -114,21 +116,40 @@ protected override void TranslateNonPublicFieldAccess(MemberExpression member)
...
-                    {
-                        Argument(Translate<ExpressionSyntax>(member.Expression)),
-                        Argument(Translate<ExpressionSyntax>(value))
-                    })));
+            AddNamespace(methodName.Namespace);
+            if (propertyInfo != null)
+            {
+                if (assignmentKind is not SyntaxKind.SimpleAssignmentExpression)
+                {
+                    throw new NotSupportedException("Compound assignment not supported");
```
Comment (ajcvickers) [inline src/EFCore.Design/Query/Internal/RuntimeModelLinqToCSharpSyntaxTranslator.cs:130]: Hard-coded string?

File: test/EFCore.Design.Tests/Query/LinqToCSharpSyntaxTranslatorTest.cs [hunk 3]
```diff
@@ -114,46 +114,45 @@ public void Private_instance_field_SimpleAssign()
...
+    [InlineData(ExpressionType.MultiplyAssign, "*=")]
+    [InlineData(ExpressionType.DivideAssign, "/=")]
+    [InlineData(ExpressionType.ModuloAssign, "%=")]
+    [InlineData(ExpressionType.SubtractAssign, "-=")]
+    [InlineData(ExpressionType.AndAssign, "&=")]
+    [InlineData(ExpressionType.OrAssign, "|=")]
+    [InlineData(ExpressionType.LeftShiftAssign, "<<=")]
+    [InlineData(ExpressionType.RightShiftAssign, ">>=")]
+    [InlineData(ExpressionType.ExclusiveOrAssign, "^=")]
     public void Private_instance_field_AssignOperators(ExpressionType expressionType, string op)
```
Comment (ajcvickers) [inline test/EFCore.Design.Tests/Query/LinqToCSharpSyntaxTranslatorTest.cs:127]: Looks like "op" is not used here.

File: test/EFCore.InMemory.FunctionalTests/Scaffolding/Baselines/ComplexTypes/PrincipalBaseEntityType.cs [hunk 4]
```diff
@@ -511,23 +477,23 @@ public static RuntimeEntityType Create(RuntimeModel model, RuntimeEntityType bas
...
+                (InternalEntityEntry entry) => UnsafeAccessor_Microsoft_EntityFrameworkCore_Scaffolding_PrincipalBase_RefTypeList((CompiledModelTestBase.PrincipalBase)entry.Entity),
+                (InternalEntityEntry entry) => UnsafeAccessor_Microsoft_EntityFrameworkCore_Scaffolding_PrincipalBase_RefTypeList((CompiledModelTestBase.PrincipalBase)entry.Entity),
+                (InternalEntityEntry entry) => entry.ReadOriginalValue<List<IPAddress>>(refTypeList, 9),
                 (InternalEntityEntry entry) => entry.GetCurrentValue<List<IPAddress>>(refTypeList),
-                (ValueBuffer valueBuffer) => valueBuffer[10]);
+                (ValueBuffer valueBuffer) => valueBuffer[9]);
             refTypeList.SetPropertyIndexes(
-                index: 10,
-                originalValueIndex: 10,
+                index: 9,
```
Comment (ajcvickers) [inline test/EFCore.InMemory.FunctionalTests/Scaffolding/Baselines/ComplexTypes/PrincipalBaseEntityType.cs:495]: Why did the property indexes change in all these tests?

**Review by ajcvickers — APPROVED** [review]

File: src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs:425]: I just made it consistent with what query generates. The idea is to have a unique name that still means something. If Shay agrees we can just use the member name + uniquefier to keep it short.

File: test/EFCore.Design.Tests/Query/LinqToCSharpSyntaxTranslatorTest.cs (hunk 3 above)
Comment (AndriySvyryd) [reply test/EFCore.Design.Tests/Query/LinqToCSharpSyntaxTranslatorTest.cs:127]: Shay's PR will take care of this

File: test/EFCore.InMemory.FunctionalTests/Scaffolding/Baselines/ComplexTypes/PrincipalBaseEntityType.cs (hunk 4 above)
Comment (AndriySvyryd) [reply test/EFCore.InMemory.FunctionalTests/Scaffolding/Baselines/ComplexTypes/PrincipalBaseEntityType.cs:495]: I made `FlagsEnum2` private, but didn't configure it explicitly in this test

File: src/EFCore.Design/Query/Internal/RuntimeModelLinqToCSharpSyntaxTranslator.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Query/Internal/RuntimeModelLinqToCSharpSyntaxTranslator.cs:130]: This is from Shay's PR, I was reading this as "not implemented", we could handle compound assignments by calling the getter and then the setter

File: src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs:425]: Sure, am fine with whatever - remember this is just generated code so nothing's very important here (except for uniqueness).

**Pull request merged**